### ISSUE TITLE
log_errors will still produce an error log when using a try/tryCatch

### ISSUE THREAD

**Issue by nathaneastwood** [daroczig/logger#39]
When I am using `logger::log_errors()`, if I want to catch failures, `logger` will still log those errors; which I don't necessarily always want. For example:

```
library(logger)
log_errors()

my_fun <- function() {
  res <- try(stop("hello world"), silent = TRUE)
  if (methods::is(res, "try-error")) {
    TRUE
  } else {
    FALSE
  }
}

my_fun()
# ERROR [2019-11-11 15:47:08] hello world
# [1] TRUE
```

**Comment by r2evans** [comment]
I'm not the dev/maintainer, but I suspect that that is because logging is done before any signal handler is taken into consideration. That is, I *believe* (I am not a `.signalCondition` guru) that `tryCatch` (and therefore `try`) intercepts the *results* of `stop`, not the *call to `stop` itself* (which is where the logging is occurring). The code within `stop` has no idea that it is going to be caught by something else. When `stop` exits, only then is when `tryCatch` determines that a signal is being raised from inside some expression ... the call to `stop` has already come and gone.

As much as I might like the behavior you suggest (*at times*), I think this might be too deep with R's internals for a user-space package to have much affect on it.

**Comment by r2evans** [comment]
The more I play with it, the more I see this is problematic in that it causes a lot of false-positives. For instance, after `log_errors()`, I often see:

```r
devtools::document("~/Projects/mypkg")
# Updating mypkg documentation
# myhost/11792/base/eval 2019-11-22 16:16:20 INFO: Updating mypkg documentation
# Writing NAMESPACE
# Loading mypkg
# myhost/11792/base/eval 2019-11-22 16:16:21 INFO: Loading mypkg
# myhost/11792/base/eval 2019-11-22 16:16:21 ERROR: not a namespace
# myhost/11792/base/eval 2019-11-22 16:16:22 ERROR: not a namespace
# Writing NAMESPACE
```

Those two errors are from somewhere in `devtools`-land (I believe), and since I've never seen them before, I assume that it's something otherwise innocuous that is wrapped with `tryCatch` and squashed/handled. While I'm really excited to be able to use something like this, I think the false-positives will either (a) cause me to troubleshoot an error that isn't mine (if I don't have namespace/function in the error message, or (b) numb me and I'll miss when something *real* happens.

Just a revisit to the topic, thanks again.

**Comment by nathaneastwood** [comment]
I've also had false-positives and they also sent me down a bit of a rabbit hole initially.

**Comment by r2evans** [comment]
@nathaneastwood, on a bit of research, I have found no way to work around this. Specifically, catching/muffling warnings and errors is extrinsic to the functions themselves; that is, `stop` doesn't know it will be censored.

It might be feasible to try to `trace` the `tryCatch`/`try` functions as well, but: (1) we'd need to trace `stop` for base stuff, and then (2) *not* re-log call to `stop` within `tryCatch` if one of the condition handlers were used. And frankly (3) this is getting way too complicated, something is bound to go wrong as we put fingers in so many different pies.

As excited as I am about the potential for converting all messages/warnings/errors into "proper" log events, I don't know that it can happen safely/easily without significant R-internal knowledge, access, even changes.

**Comment by nathaneastwood** [comment]
I have looked into the possibility of writing my own `stop()` function by calling the appropriate source code but I need to do more research since I don't develop in C. I think this could be a way forward as adding `trace` to `try`/`tryCatch` is starting to get messy.

**Comment by jozefhajnala** [comment]
One could change the `tracer`  within the `log_errors()` call to `trace()` to get customized behavior. For instance, deparsing the `sys.calls()` and looking for `try` or `tryCatch`. Not very elegant, but this could be a direction:

```r
log_errors <- function() {
    invisible(suppressMessages(trace(
        what = 'stop',
        tracer = substitute({
            firstCall <- deparse(sys.calls()[[1L]])
            if (isTRUE(grepl("^try\\(|^tryCatch\\(", firstCall[[1L]]))) {
              logger::log_warn(logger::skip_formatter(paste(c(firstCall, " - try call failed: ", list(...)), collapse = '')))
            } else {
              logger::log_error(logger::skip_formatter(paste(list(...), collapse = '')))
            }
        }),
        print = FALSE,
        where = baseenv())))
}
```

That would give for this code:

```r
fun_wrapper <- function() fun_with_error()
fun_with_error <- function() stop("Stopped")
logger::log_errors()
try({1 + 1; fun_wrapper(); 2}, silent = TRUE)
```

This output:

```
WARN [2019-11-26 15:17:01] try({    1 + 1    fun_wrapper()    2}, silent = TRUE) - try call failed: Stopped
```

I could draft a (more polished) PR if @daroczig would consider this direction.

**Comment by r2evans** [comment]
@nathaneastwood, while formidable and commendable, I think that writing a new `stop()` defeats a convenience of `log_errors()`: the ability to log errors in existing code (local or from other packages) without modification.

@jozefhajnala, interesting approach. While I think finding no evidence of `try` or `tryCatch` is evidence that the error is not likely to be suppressed in normal user code, there are some holes in the logic:

1. it's feasible to `tryCatch`, check the error, and pass it on (meaning do *not* suppress it), so presence of this function is insufficient; I use this periodically to catch an error I know about but not suppress errors I'm not expecting;

2. can't errors also be caught with `withCallingHandlers` and/or similar lower-level functions? Or is that just for `warning`s?

**Comment by nljhaj1** [comment]
@r2evans I agree, where I was heading with this is that the desired behavior is very context-dependent, and what we are trying to decide here is a behavior that works for "most cases". You might actually want to see exactly what the function currently does, or maybe you would want to see something else based on the context in which the error occurred (one example is not showing ERROR when in a `tryCatch` block).

We could make the behavior I outlined more customizable, e.g. by
- exposing the condition that I hard-coded in the example to a naive `grepl("^try\\(|^tryCatch\\(", firstCall[[1L]])` via an option or an environment variable (or something else)
- exposing the behavior once the condition is met that I hard-coded in my example to `logger::log_warn(...)` once again via an option or an environment variable (or something else)
- multiple conditions and behaviors
- provide other ways to examine the context of the error (e.g. when passing them on, assigning those a class and write class-based methods)

But in the end I believe some user "decision" will be desired in any case, as this to me is customizing a behavior that is context and purpose dependent.

**Comment by jozefhajnala** [comment]
@r2evans I agree, where I was heading with this is that the desired behavior is very context-dependent, and what we are trying to decide here is a behavior that works for "most cases". You might actually want to see exactly what the function currently does, or maybe you would want to see something else based on the context in which the error occurred (one example is not showing ERROR when in a `tryCatch` block).

We could make the behavior I outlined more customizable, e.g. by

- exposing the condition that I hard-coded in the example to a naive `grepl("^try\\(|^tryCatch\\(", firstCall[[1L]])` via an option or an environment variable (or something else)
- exposing the behavior once the condition is met that I hard-coded in my example to `logger::log_warn(...)` once again via an option or an environment variable (or something else)
- multiple conditions and behaviors
- provide other ways to examine the context of the error (e.g. when passing them on, assigning those a class and write class-based methods)
- other ways that did not come to mind ;-)

But in the end I believe some user "decision" will be desired in any case, as this to me is customizing a behavior that is context and purpose dependent.

**Comment by r2evans** [comment]
So true. Your premise of making it a user-selectable behavior is spot-on in my opinion.

I was initially thinking of an argument to `log_errors`, but your suggestion of `options` or envvars allows modifying this behavior without the need for `untrace(stop);log_errors()`.

I also think that the ideal of "only log uncaught errors" can only be done with modifications to internal R code, so is unlikely ... so the user needs to accept either false-positives (logging errors that have been caught and suppressed) or false-negatives (not logging an error just because there was a `tryCatch` somewhere in the chain, regardless of what it did). Clarity and transparency in the pros/cons is key here.

While it is getting a little more complicated, do you think adding a `trace` to `tryCatch` could work? For instance, add an `exit` clause that determines if `returnValue()` contains an error and log it there, otherwise don't log it? That would include `try` (which is just a wrapper) as well, though would miss more-complicated approaches that directly deal with handlers (including `withCallingHandlers` and `.Internal(.addCondHands(...))`). (On initial testing, perhaps this is a bad idea ... it can produce a cycle of intercepted calls ...)

**Comment by daroczig** [comment]
Thanks, everyone -- these are all valid points, and actually I'm really not happy with the current implementation, but wanted to keep the behaviour similar to warnings / messages.

The main problem with `log_errors` I think is that

* it's an ugly hack
* even handled `stop` calls will be captured
* errors outside of `stop` are not captured, eg syntax error on the top level

I was playing with setting the `error` option, which works much nicer:

```r
library(logger)
options(error = function() {
    log_error(geterrmessage())
    if (!interactive()) {
        quit('no', status = 1, runLast = FALSE)
    }
})

f <- function(x, y = 2) x*y

for (i in list(1, 'foobar', mean)) {
    devnull <- tryCatch(f(i), error = function(e) NA_real_)
}

devnull <- try(f(i), silent = TRUE)

devnull <- f(i)
```

But then it's not consistent with the warnings / messages handlers and has some other issues.

In short, I actually gave up on using `log_errors` as all my scripts simply stops at the first error anyway and prints the error message (that is already monitored) .. so not much gain with the automatic error logger -- except for adding extra timestamp and potentially delivering to non-standard destination.

Anyway, thoughts on the above `options` approach?

**Comment by r2evans** [comment]
Actually, I think the `options(error=...)` approach is perfectly acceptable for errors. (The only thing I'd add would be explicitly use `logger::log_error`.) What other issues have you found with this technique?

I was really excited to suggest using `.Internal(.addCondHands(...))` for messages and warnings, but it appears that any error condition resets all others condition handlers:

```r
mhandler <- function(m) { logger::log_info(conditionMessage(m)); m; }
whandler <- function(w) { logger::log_warn(conditionMessage(w)); w; }
.Internal(.addCondHands(c("message", "warning"),
                        list(message=mhandler, warning=whandler),
                        globalenv(), NULL, TRUE))
message("quux")
# INFO [2019-11-27 09:24:08] quux
# quux
suppressMessages(message("quux"))
message("quux")
# INFO [2019-11-27 09:24:10] quux
# quux

a
# Error: object 'a' not found
message("quux")
# quux
```

If you can think of a way to bypass that, then this might be a viable alternative. I tried adding that `.Internal(...)` code to the `options(error=...)` call, but that did not work. (My guess is that the condition-handler cleanup might be later in the stack.)

**Comment by tdeenes** [comment]
`globalCallingHandlers` in R 4.0.0. will make it easy to auto-log errors in a safe way (without affecting errors which are explicitly handled in tryCatch etc.):
https://developer.r-project.org/blosxom.cgi/R-devel/NEWS/2020/03/22#n2020-03-22

**Comment by tdeenes** [comment]
```r
globalCallingHandlers(
  warning = function(w) print("Huh, I caught a warning"), 
  error = function(e) print("Houston, we have a problem")
)
log(-1)
# [1] "Huh, I caught a warning"
# [1] NaN
# Warning message:
# In log(-1) : NaNs produced

log(something)
# [1] "Houston, we have a problem"
# Error: object 'something' not found

tryCatch(something, error = function(e) print("Issue solved"))
# [1] "Issue solved"
```

**Comment by r2evans** [comment]
Looks very promising! I should probably upgrade to 3.6 soon so I'm not as far behind when 4.0 comes out ...

**Comment by daroczig** [comment]
Thanks, folks -- just pushed an update that should solve this. I'm still adding some unit tests, but I think overall it's in a good shape. Please reopen if you have any further suggestions.
